x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportDeleteJobAction.java

```diff
@@ -328,72 +330,97 @@ private void deleteJobDocuments(ParentTaskAssigningClient parentTaskClient, Stri
-        // If we do not have any hits, we can drop the indices and then skip the DBQ and alias deletion.
-        ActionListener<SearchResponse> customIndexSearchHandler = ActionListener.wrap(
-                searchResponse -> {
-                    if (searchResponse == null || searchResponse.getHits().getTotalHits().value > 0) {
-                        deleteByQueryExecutor.onResponse(true); // We need to run DBQ and alias deletion
-                    } else {
-                        logger.info("Running DELETE Index on [" + String.join(", ", indexNames.get()) + "] for job [" + jobId + "]");
-                        DeleteIndexRequest request = new DeleteIndexRequest(indexNames.get());
-                        request.indicesOptions(IndicesOptions.lenientExpandOpen());
-                        // If we have deleted the index, then we don't need to delete the aliases or run the DBQ
-                        executeAsyncWithOrigin(
-                                parentTaskClient.threadPool().getThreadContext(),
-                                ML_ORIGIN,
-                                request,
-                                ActionListener.<AcknowledgedResponse>wrap(
-                                        response -> deleteByQueryExecutor.onResponse(false), // skip DBQ && Alias
-                                        failureHandler),
-                                parentTaskClient.admin().indices()::delete);
-                    }
-                },
-                failure -> {
-                    if (ExceptionsHelper.unwrapCause(failure) instanceof IndexNotFoundException) { // assume the index is already deleted
-                        deleteByQueryExecutor.onResponse(false); // skip DBQ && Alias
-                    } else {
-                        failureHandler.accept(failure);
-                    }
-                }
+        // Step 6. Handle each multi-search response. There should be one response for each underlying index.
+        // For each underlying index that contains results ONLY for the current job, we will delete that index.
+        // If there exists at least 1 index that has another job's results, we will run DBQ.
+        ActionListener<MultiSearchResponse> customIndexSearchHandler = ActionListener.wrap(
+           multiSearchResponse -> {
+               if (multiSearchResponse == null) {
+                   deleteByQueryExecutor.onResponse(true); // We need to run DBQ and alias deletion
+                   return;
+               }
+               String defaultSharedIndex = AnomalyDetectorsIndexFields.RESULTS_INDEX_PREFIX +
+                   AnomalyDetectorsIndexFields.RESULTS_INDEX_DEFAULT;
+               List<String> indicesToDelete = new ArrayList<>();
+               boolean needToRunDBQTemp = false;
+               assert multiSearchResponse.getResponses().length == indexNames.get().length;
```

Comment:
I'm not sure how it looks like in the rest of our codebase but I think `assert` should only be used to assert on the invariants implied directly by the code. 
In this case we examine a response from an external service which can have a bug and therefore cause the assertion to fail. I don't know how such an exception would behave.
WDYT?

Reply:
It could be a bug in the underlying service. OR it could be a bug in that we did not provide the correct number of clauses ("invariants implied directly by the code").

Other places in our code base have similar assertions.